src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -262,34 +266,31 @@ internal sealed class RSession : IRSession, IRCallbacks {
 
             await _initializationTcs.Task;
             initializationCts.Dispose();
+            _stopHostLock.EnqueueReset();
             lockToken.Set();
         }
 
-        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
-        public IRSessionReconnectTransaction StartReconnecting() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionReconnectTransaction StartReconnecting() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
         public async Task StopHostAsync() {
             _disposeToken.ThrowIfDisposed();
```

Comment:
Cancellation looks OK. However, how do we wait for the cancellation to actually complete and do we need to? We do have some code that, say, requests cancellation and then assumes it is done. But it may task some time to actually throw, right?

Cancellation before dispose also matches state machine - imagine requests come and get placed in a queue to get processed and there is a worker that processes them. So Start -> Dispose would be handled as `Start` -> `Cancel (if worker still running)` -> `Wait for Cancellation` -> `Dispose` or `Stop` (if already started) -> `Wait for Stopped` and then `Dispose`.